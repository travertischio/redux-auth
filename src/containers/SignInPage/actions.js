/*
 *
 * SignInPage actions
 *
 */

import {
  SIGN_IN_ACTION,
  SIGN_IN_SUCCEED_ACTION,
  SIGN_IN_FAILED_ACTION,
  DESTROY_PAGE_ACTION,
} from './constants';

export function signInAction(credentials) {
  return {
    type: SIGN_IN_ACTION,
    payload: credentials,
  };
}

export function signInSucceedAction(response) {
  return {
    type: SIGN_IN_SUCCEED_ACTION,
    payload: response,
  };
}

export function signInFailedAction(rejection) {
  return {
    type: SIGN_IN_FAILED_ACTION,
    payload: rejection,
  };
}

export function destroyPageAction() {
  return {
    type: DESTROY_PAGE_ACTION,
  };
}
