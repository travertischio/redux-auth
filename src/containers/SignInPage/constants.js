/*
 *
 * SignInPage constants
 *
 */

export const SIGN_IN_ACTION = 'app/SignInPage/SIGN_IN';
export const SIGN_IN_SUCCEED_ACTION = 'app/SignInPage/SUCCEED';
export const SIGN_IN_FAILED_ACTION = 'app/SignInPage/FAILED';
export const DESTROY_PAGE_ACTION = 'app/SignInPage/DESTROY';
