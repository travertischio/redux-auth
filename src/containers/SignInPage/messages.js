/*
 * SignInPage Messages
 *
 * This contains all the text for the SignInPage component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  pageTitle: {
    id: 'app.containers.SignInPage.pageTitle',
    defaultMessage: 'Sign In',
  },
  pageDescription: {
    id: 'app.containers.SignInPage.pageDescription',
    defaultMessage: 'Sign In',
  },
  header: {
    id: 'app.containers.SignInPage.header',
    defaultMessage: 'This is SignInPage container !',
  },
});
