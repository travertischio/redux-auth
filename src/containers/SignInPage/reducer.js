/*
 *
 * SignInPage reducer
 *
 */

import { fromJS } from 'immutable';
import _isArray from 'lodash/isArray';
import {
  BLOCKED_ACCOUNT_ACTION,
  REQUIRE_CAPTCHA_ACTION,
} from '~/containers/AuthenticationProvider/constants';
import {
  DESTROY_PAGE_ACTION,
  SIGN_IN_ACTION,
  SIGN_IN_FAILED_ACTION,
  SIGN_IN_SUCCESS_ACTION,
} from './constants';

const initialState = fromJS({});

function signInPageReducer(state = initialState, action) {
  switch (action.type) {
    case SIGN_IN_ACTION:
      return onSignInAction(state);
    case SIGN_IN_SUCCESS_ACTION:
      return onSignSuccessAction(state);
    case SIGN_IN_FAILED_ACTION:
      return onSignFailedAction(state, action.payload);
    case DESTROY_PAGE_ACTION:
      return onDestroyPageAction(state);
    case BLOCKED_ACCOUNT_ACTION:
      return onBlockedAccountAction(state);
    case REQUIRE_CAPTCHA_ACTION:
      return onRequireCaptchaAction(state);
    default:
      return state;
  }
}

function onSignInAction(state) {
  return state.merge({
    loading: true,
    errorMessage: null,
    captchaRequired: false,
  });
}

function onSignSuccessAction(state) {
  return state.merge({
    loading: false,
    errorMessage: null,
  });
}

function onSignFailedAction(state, rejection) {
  const { response } = rejection;
  let errorMessage = 'Unable to sign in. Please try again.';

  if (response && response.status === 400) {
    const { nonFieldErrors } = response.data;

    if (_isArray(nonFieldErrors)) {
      errorMessage = nonFieldErrors.join(', ');
    }
  }

  return state.merge({
    loading: false,
    errorMessage,
  });
}

function onDestroyPageAction() {
  return fromJS({});
}

function onBlockedAccountAction(state) {
  return state.set('blockedAccount', true);
}

function onRequireCaptchaAction(state) {
  return state.set('captchaRequired', true);
}

export default signInPageReducer;
