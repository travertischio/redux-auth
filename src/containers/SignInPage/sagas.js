import {
  call,
  cancel,
  put,
  select,
  take,
  takeEvery,
  takeLatest,
} from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { signIn as signInApiCall } from '~/api';
import { handleAuthenticationSaga } from '~/containers/AuthenticationProvider/sagas';
import { makeSelectLastUserToken } from '~/containers/AuthenticationProvider/selectors';
import {
  signInFailedAction,
  signInSuccessAction,
} from './actions';
import {
  SIGN_IN_ACTION,
  SIGN_IN_SUCCESS_ACTION,
} from './constants';

export function* watchSignInAction() {
  const signInActionWatcher = yield takeLatest(SIGN_IN_ACTION, signInSaga);

  yield take(LOCATION_CHANGE);
  yield cancel(signInActionWatcher);
}

export function* signInSaga(action) {
  try {
    const selectLastUserToken = yield call(makeSelectLastUserToken, action.credentials.get('email'));
    const lastToken = yield select(selectLastUserToken);
    const credentials = {
      token: lastToken,
      ...action.credentials.toJS(),
    };

    const response = yield call(signInApiCall, credentials);

    yield put(signInSuccessAction(response));
  } catch (error) {
    yield put(signInFailedAction(error));
  }
}

export function* watchSignInSuccessAction() {
  const signInSuccessActionWatcher = yield takeEvery(SIGN_IN_SUCCESS_ACTION, handleAuthenticationSaga);

  yield take(LOCATION_CHANGE);
  yield cancel(signInSuccessActionWatcher);
}

export default [
  watchSignInAction,
  watchSignInSuccessAction,
];
