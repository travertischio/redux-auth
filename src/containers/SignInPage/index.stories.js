import React from 'react';
import { storiesOf } from '@kadira/storybook';
import SignInPage from '.';

storiesOf('Sign In Page', module)
  .add('Sign In Page', () => <SignInPage />);
