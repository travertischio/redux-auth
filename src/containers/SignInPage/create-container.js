import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import { injectIntl } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import config from '~/config';
import UserIsNotAuthenticated from '~/hocs/AuthWrappers/UserIsNotAuthenticated';
import selectSignInPage from './selectors';
import messages from './messages';
import {
  signInAction,
  destroyPageAction,
} from './actions';

const AuthWrapper = config.signInAuthWrapper || UserIsNotAuthenticated;

export default function createSignInContainer(PageComponent, options = {}) {
  const mapStateToProps = createStructuredSelector({
    SignInPage: selectSignInPage,
  });

  const mapDispatchToProps = {
    onSubmitForm: signInAction,
    onUnMount: destroyPageAction,
  };

  @AuthWrapper
  @injectIntl
  @connect(mapStateToProps, mapDispatchToProps)
  class SignInContainer extends PureComponent {
    static propTypes = {
      intl: PropTypes.object.isRequired,
      onUnMount: PropTypes.func.isRequired,
    };

    componentWillUnmount() {
      this.props.onUnMount();
    }

    render() {
      const { formatMessage } = this.props.intl;
      const pageTitle = formatMessage(messages.pageTitle);
      const pageDescription = formatMessage(messages.pageDescription);

      return (
        <div className={options.className || 'sign-in-page'}>
          {!options.noHelmet &&
            <Helmet
              title={pageTitle}
              meta={[
                { name: 'description', content: pageDescription },
              ]}
            />
          }

          <PageComponent {...this.props} />
        </div>
      );
    }
  }

  return SignInContainer;
}
