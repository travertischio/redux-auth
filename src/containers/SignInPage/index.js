/*
 *
 * SignInPage
 *
 */

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import { injectIntl } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import SignInForm from '../../components/SignInForm';
import UserIsNotAuthenticated from '../../hocs/AuthWrappers/UserIsNotAuthenticated';
import makeSelectSignInPage from './selectors';
import messages from './messages';
import { signInAction, destroyPageAction } from './actions';

const mapStateToProps = createStructuredSelector({
  SignInPage: makeSelectSignInPage(),
});

const mapDispatchToProps = (dispatch) => ({
  onSubmitForm: (values) => dispatch(signInAction(values)),
  onUnMount: () => dispatch(destroyPageAction()),
});

@UserIsNotAuthenticated
@injectIntl
@connect(mapStateToProps, mapDispatchToProps)
export default class SignInPage extends PureComponent { // eslint-disable-line react/prefer-stateless-function

  componentWillUnmount() {
    this.props.onUnMount();
  }

  render() {
    const {
      onSubmitForm,
      SignInPage: {
        loading,
        errorMessage,
      },
    } = this.props;
    const { formatMessage } = this.props.intl;
    const pageTitle = formatMessage(messages.pageTitle);
    const pageDescription = formatMessage(messages.pageDescription);

    return (
      <div>
        <Helmet
          title={pageTitle}
          meta={[
            { name: 'description', content: pageDescription },
          ]}
        />

        { loading && <div>Processing... Please wait.</div> }
        { errorMessage && <div>{errorMessage}</div> }

        <SignInForm onSubmit={onSubmitForm} />
      </div>
    );
  }
}

SignInPage.propTypes = {
  intl: PropTypes.object,
  SignInPage: PropTypes.object,
  onSubmitForm: PropTypes.func,
  onUnMount: PropTypes.func,
};
