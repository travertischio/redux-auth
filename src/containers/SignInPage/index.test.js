import React from 'react';
import {
  getStoreWithInitialState,
  findActionByType,
  mountWithIntl,
  createComponentWithRouter,
} from 'react-unit-testing-utils';
import SignInPage from './index';
import { destroyPageAction } from './actions';
import {
  SIGN_IN_ACTION,
  DESTROY_PAGE_ACTION,
} from './constants';

describe('<SignInPage />', () => {
  it('should dispatch action SIGN_IN_ACTION when fill and submit the form', () => {
    const store = getStoreWithInitialState({
      auth: {
        isAuthenticated: false,
      },
      signInPage: {},
    });

    const wrapper = mountWithIntl(<SignInPage />, {
      store,
    });

    const email = 'tester@test.com';
    const password = 'zxy123';

    wrapper.find('input[name="email"]').simulate('change', { target: { value: email } });
    wrapper.find('input[name="password"]').simulate('change', { target: { value: password } });
    wrapper.find('form').simulate('submit');

    const recivedAction = findActionByType(store, SIGN_IN_ACTION);

    expect(recivedAction.type).toEqual(SIGN_IN_ACTION);
  });

  it('should dispatch action DESTROY_PAGE_ACTION when the component unmount', () => {
    const initialState = {
      auth: {
        isAuthenticated: false,
      },
      signInPage: {},
    };
    const { wrapper, store } = createComponentWithRouter(<SignInPage />, initialState);
    wrapper.unmount();
    const recivedAction = findActionByType(store, DESTROY_PAGE_ACTION);
    expect(recivedAction).toEqual(destroyPageAction());
  });

  it('should NOT render SignInPage when user IS authenticated', () => {
    const initialState = {
      auth: {
        isAuthenticated: true,
        user: {},
      },
    };
    const { wrapper } = createComponentWithRouter(<SignInPage />, initialState);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render SignInPage when user IS NOT authenticated', () => {
    const initialState = {
      auth: {
        isAuthenticated: false,
      },
      signInPage: {},
    };
    const { wrapper } = createComponentWithRouter(<SignInPage />, initialState);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render SignInPage with loading indicator when loading is in the props', () => {
    const initialState = {
      auth: {
        isAuthenticated: false,
      },
      signInPage: {
        loading: true,
      },
    };
    const { wrapper } = createComponentWithRouter(<SignInPage />, initialState);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render SignInPage with error message when errorMessage is in the props', () => {
    const initialState = {
      auth: {
        isAuthenticated: false,
      },
      signInPage: {
        errorMessage: 'Unable to sign in. Please try again.',
      },
    };
    const { wrapper } = createComponentWithRouter(<SignInPage />, initialState);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
