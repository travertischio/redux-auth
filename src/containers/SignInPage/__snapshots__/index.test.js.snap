// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignInPage /> should NOT render SignInPage when user IS authenticated 1`] = `<main />`;

exports[`<SignInPage /> should render SignInPage when user IS NOT authenticated 1`] = `
<main>
  <div
    className="sign-in-page"
  >
    <div>
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <div>
          <label
            htmlFor="email"
          >
            E-mail
          </label>
          <div
            className=""
          >
            <input
              id="email"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="E-mail"
              type="email"
              value=""
            />
          </div>
        </div>
        <div>
          <label
            htmlFor="password"
          >
            Password
          </label>
          <div
            className=""
          >
            <input
              id="password"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <button
            disabled={true}
            type="submit"
          >
            Sign In
          </button>
        </div>
      </form>
    </div>
  </div>
</main>
`;

exports[`<SignInPage /> should render SignInPage with error message when errorMessage is in the props 1`] = `
<main>
  <div
    className="sign-in-page"
  >
    <div>
      <div>
        Unable to sign in. Please try again.
      </div>
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <div>
          <label
            htmlFor="email"
          >
            E-mail
          </label>
          <div
            className=""
          >
            <input
              id="email"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="E-mail"
              type="email"
              value=""
            />
          </div>
        </div>
        <div>
          <label
            htmlFor="password"
          >
            Password
          </label>
          <div
            className=""
          >
            <input
              id="password"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <button
            disabled={true}
            type="submit"
          >
            Sign In
          </button>
        </div>
      </form>
    </div>
  </div>
</main>
`;

exports[`<SignInPage /> should render SignInPage with loading indicator when loading is in the props 1`] = `
<main>
  <div
    className="sign-in-page"
  >
    <div>
      <div>
        Processing... Please wait.
      </div>
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <div>
          <label
            htmlFor="email"
          >
            E-mail
          </label>
          <div
            className=""
          >
            <input
              id="email"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="E-mail"
              type="email"
              value=""
            />
          </div>
        </div>
        <div>
          <label
            htmlFor="password"
          >
            Password
          </label>
          <div
            className=""
          >
            <input
              id="password"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <button
            disabled={true}
            type="submit"
          >
            Sign In
          </button>
        </div>
      </form>
    </div>
  </div>
</main>
`;
